### PR TITLE
Instrument query path

### PR DESCRIPTION
**Pull request by ryanhall07**
Add additional metrics to better understand the bottlenecks in the query
path.

**What this PR does / why we need it**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/dbnode/network/server/tchannelthrift/node/service.go [hunk 1]
```diff
@@ -123,6 +123,24 @@ type serviceMetrics struct {
...
+}
+
+func newQueryMetrics(name string, scope tally.Scope) queryMetrics {
+	return queryMetrics{
+		byRange: limits.NewQueryRangeMetrics(name, scope),
+		byDocs:  limits.NewCardinalityMetrics(name, scope),
+	}
+}
+
+type queryMetrics struct {
```
Comment (wesleyk) [inline src/dbnode/network/server/tchannelthrift/node/service.go:141]: can we move this into the limits package? Seems like we typically want the combo of by range / by cardinality together

File: src/dbnode/network/server/tchannelthrift/node/service.go [hunk 2]
```diff
@@ -875,6 +904,12 @@ type fetchTaggedResultsIter struct {
 	tagEncoder      serialize.TagEncoder
 	iOpts           instrument.Options
 	instrumentClose func(error)
+	startTime       time.Time
```
Comment (wesleyk) [inline src/dbnode/network/server/tchannelthrift/node/service.go:907]: where is this set?

Comment (wesleyk) [reply src/dbnode/network/server/tchannelthrift/node/service.go:907]: also what's the difference between this and `fetchStart`?

File: src/dbnode/network/server/tchannelthrift/node/service.go [hunk 3]
```diff
@@ -826,6 +850,11 @@ func (s *service) fetchTaggedIter(ctx context.Context, req *rpc.FetchTaggedReque
 		tagEncoder:      tagEncoder,
 		iOpts:           s.opts.InstrumentOptions(),
 		instrumentClose: instrumentClose,
+		totalDocsCount:  queryResult.Results.TotalDocsCount(),
+		nowFn:           s.nowFn,
+		fetchStart:      startTime,
```
Comment (wesleyk) [inline src/dbnode/network/server/tchannelthrift/node/service.go:855]: is this supposed to include the index lookup? Looks like it does (line 835)

**Review by wesleyk — APPROVED** [review]
LGTM

File: src/dbnode/network/server/tchannelthrift/node/service.go (hunk 3 above)
Comment (ryanhall07) [reply src/dbnode/network/server/tchannelthrift/node/service.go:855]: yes. it's total time of the fetch api call

File: src/dbnode/network/server/tchannelthrift/node/service.go (hunk 2 above)
Comment (ryanhall07) [reply src/dbnode/network/server/tchannelthrift/node/service.go:907]: ugh it's not. clearly a test is needed for the metrics. 

renamed this to `dataReadStart` to make it more clear.

File: src/dbnode/network/server/tchannelthrift/node/service.go (hunk 1 above)
Comment (ryanhall07) [reply src/dbnode/network/server/tchannelthrift/node/service.go:141]: it doesn't have anything to do with limits though.

Comment (ryanhall07) [reply src/dbnode/network/server/tchannelthrift/node/service.go:141]: actually i think it's strange they are in the limits package. moved them to the `node` package for the service.

File: src/dbnode/storage/index/metrics.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+			Tagged(map[string]string{"count": value}).
+			Histogram("timing", instrument.SparseHistogramTimerHistogramBuckets()),
+	}
+}
+
+type queryCardinality struct {
+	cardinalityHists []*cardinalityHist
+}
+
+func (bm *queryCardinality) Record(seriesCount int, queryRuntime time.Duration) {
```
Comment (robskillington) [inline src/dbnode/storage/index/metrics.go:126]: This should be `docsCount` yeah? Maybe we should rename all this "cardinality" metrics naming to "queryDocsCount" metrics?

**Review by robskillington — APPROVED** [review]
LGTM

**Pull request merged**